Add @parcel/css to benchmark

onigoetz opened the pull request:
Hello,

I added [@parcel/css](https://www.npmjs.com/package/@parcel/css) as a benchmark.
It's an alpha but interestingly it's the fastest tool in every test.

Fixes #100

onigoetz committed:
Add @parcel/css to benchmark

devongovett commented on lib/minify.js:
```diff
@@ -37,6 +38,9 @@ const minifiers = {
     return esbuild.transformSync(source, { loader: 'css', minify: true }).code;
+  },
+  '@parcel/css': source => {
+    return parcelCss.transform({filename: "", code: Buffer.from(source), minify: true}).code;
```
If you want to get even smaller output, you can specify browser targets. This allows `@parcel/css` to e.g. remove vendor prefixes that aren't needed. See [here](https://github.com/parcel-bundler/parcel-css/blob/master/test.js#L11) for an example. Maybe it could be a separate column for that similar to the "advanced" modes of the other tools if you want to keep it separate. However, unlike those advanced modes, removing prefixes is a safe transform.

onigoetz replied:
Yes I copied from the example benchmark you have in the repository. I wanted the default entry in the benchmark to be similar to other tools as they don't have any browser-specific optimisation and thus I didn't know which browsers would be "reasonabe" to set.

devongovett replied:
Sure. I definitely think it should be clearer from the benchmark results that many of the smallest sizes shown are using unsafe transforms though. Maybe something to work on separately.

onigoetz commented:
Hi @XhmikosR, is there something I can help with to get this merged ?

dominikg commented on package.json:
```diff
@@ -31,6 +31,7 @@
   "main": "lib/index.js",
   "dependencies": {},
   "devDependencies": {
+    "@parcel/css": "^1.0.0-alpha.4",
```
can be updated to `    "@parcel/css": "^1.0.0",` now that it has been released

dominikg commented:
note: with 1.0.0 i got errors on some input files
```
parcel err SyntaxError: Unexpected token Delim('*')
    at Object.@parcel/css (/home/dominikg/develop/css-minification-benchmark/lib/minify.js:45:21)
    at Object.measure (/home/dominikg/develop/css-minification-benchmark/lib/minify.js:97:48)
    at measure (/home/dominikg/develop/css-minification-benchmark/lib/index.js:23:30)
    at /home/dominikg/develop/css-minification-benchmark/lib/index.js:51:34
    at _fulfilled (/home/dominikg/develop/css-minification-benchmark/node_modules/q/q.js:854:54)
    at /home/dominikg/develop/css-minification-benchmark/node_modules/q/q.js:883:30
    at Promise.promise.promiseDispatch (/home/dominikg/develop/css-minification-benchmark/node_modules/q/q.js:816:13)
    at /home/dominikg/develop/css-minification-benchmark/node_modules/q/q.js:624:44
    at runSingle (/home/dominikg/develop/css-minification-benchmark/node_modules/q/q.js:137:13)
    at flush (/home/dominikg/develop/css-minification-benchmark/node_modules/q/q.js:125:13)
```

XhmikosR commented:
Happy to include this assuming tests pass.

dominikg commented:
the errors are caused by outdated css invalid css syntax to please IE7 eg. `*zoom: 1` in some of the data files.

Maybe you could update the pool to only include modern sets @XhmikosR 
Or add error-handling in a way that it skips over and marks the error on the benchmark results (would skew total runtime results a bit tho).

XhmikosR commented:
Not sure about it, it will need time which I don't have right now, plus the other minifiers work with it even if it's invalid per the specs.

onigoetz commented:
I updated to @parcel/css 1.0.0 Indeed two files are in error, I changed the code a bit to correctly display that those are in error (the `isValid()` check didn't return `false` for the checks in error)

onigoetz commented:
Fixed the warnings from the build as well

XhmikosR commented on lib/minify.js:
```diff
@@ -37,6 +38,9 @@ const minifiers = {
     return esbuild.transformSync(source, { loader: 'css', minify: true }).code;
+  },
+  '@parcel/css': source => {
+    return parcelCss.transform({ filename: '', code: require('buffer').Buffer.from(source), minify: true }).code;
```
Please move the require at the top

onigoetz replied:
sure, done

XhmikosR commented:
I'm not sure the `error` changes are right. It seems it affects the total values too?

onigoetz commented:
It does but not too badly : 

```
║ TOTAL - 3462799 bytes              │ 2752429 bytes - 1862.21 ms │ 2737446 bytes - 2287.01 ms  │ 2720884 bytes - 6105.36 ms │ 2509596 bytes - 5797.12 ms │ 2447799 bytes - 1514.23 ms │ 2755245 bytes - 625.9 ms       │ 2764893 bytes - 303.51 ms │ errorerror
```

It just concatenates the "error" labels, but still excludes the minifier that has at least one error

before it was printing something like `NaNundefinederrorundefinederrorundefinedundefinedundefined`

XhmikosR commented:
We need this solution to be more robust, although I don't have a suggestion right now...

I'll just remove the offending CSS files.

XhmikosR committed:
Remove old data files which are causing issues with @percel/css

XhmikosR committed:
Update results

XhmikosR commented:
Thanks! If you have time to look at any other issues or improve the error situation, feel free to CC me (I don't watch repositories 'cause I'd go totally crazy 😛)

XhmikosR commented:
And BTW, @parcel/css is super fast, almost unbelievable! I need to try it out.

onigoetz commented:
Yes I'll have a look into this, I'm working on a postcss plugin for @parcel/css right now and I'll come back to this after.

No worries, your mental health is more important than open source projects ;-)
